cow_instrument/PointPathComponent.h

```diff
@@ -0,0 +1,111 @@
+#ifndef POINT_PATH_COMPONENT_H
+#define POINT_PATH_COMPONENT_H
+
+#include "PathComponent.h"
+#include "V3D.h"
+#include "Component.h"
+#include "IdType.h"
+
+/**
+ * CRTP type to do boiler plate work of creating PathComponents that behave like
+ * points from a neutronic perspective.
+ */
+template <typename T> class PointPathComponent : public PathComponent {
+
+public:
+  PointPathComponent(V3D pos, ComponentIdType id);
+  virtual V3D getPos() const override;
+  virtual void deltaPos(const V3D &pos) override;
+
+  virtual bool equals(const Component &other) const override;
+  virtual void registerContents(
+      std::vector<const Detector *> &detectorLookup,
+      std::vector<const PathComponent *> &pathLookup) const override;
+  virtual ComponentIdType componentId() const override;
+  virtual std::string name() const override;
+
+  double length() const override;
+  V3D entryPoint() const override;
+  V3D exitPoint() const override;
+  virtual ~PointPathComponent() {}
+  virtual PointPathComponent<T> *clone() const override;
+
+  bool operator==(const PointPathComponent<T> &other) const;
+  bool operator!=(const PointPathComponent<T> &other) const;
+
+private:
+  V3D m_pos;
+  ComponentIdType m_componentId;
+};
+
+template <typename T>
+PointPathComponent<T>::PointPathComponent(V3D pos, ComponentIdType id)
+    : m_pos(pos), m_componentId(id) {}
+
+template <typename T> V3D PointPathComponent<T>::getPos() const {
+  return m_pos;
+}
+
+template <typename T> void PointPathComponent<T>::deltaPos(const V3D &pos) {
+  m_pos[0] += pos[0];
+  m_pos[1] += pos[1];
+  m_pos[2] += pos[2];
+}
+
+template <typename T>
+PointPathComponent<T> *PointPathComponent<T>::clone() const {
+  return new PointPathComponent<T>(*this);
```

Comment:
So I think we are talking about different parts of that function here. Yes I could change the internals of the clone. Here's a toy example:

```cpp
class Abstract {
  public:
    virtual Abstract* clone() const = 0;
};

template
<typename T>
class Base : public Abstract{
public:
    virtual Base<T>* clone() const{
        return new T(*static_cast<const T*>(this));
    }
    virtual ~Base(){}

};

class Derived : public Base<Derived>{
public:
    virtual ~Derived(){}
};
```
However, this (what I was talking about) is not legal, which means that the return type is not fully covariant down to the most derived type:

```cpp
class Abstract {
  public:
    virtual Abstract* clone() const = 0;
};

template
<typename T>
class Base : public Abstract{
public:
    virtual T* clone() const{
        return new T(*static_cast<const T*>(this));
    }
    virtual ~Base(){}

};

class Derived : public Base<Derived>{
public:
    virtual ~Derived(){}
};
```